File: apps/web/lib/api/prompt-edit.ts
import { apiFetch } from "./client";
import type {
  PromptEditApplyRequest,
  PromptEditApplyResponse,
  PromptEditInstructRequest,
  PromptEditInstructResponse,
  PromptEditRegenerateRequest,
  PromptEditSessionListResponse,
} from "./types";

export async function listPromptEditSessions(
  projectId: string,
  chapterId: string,
): Promise<PromptEditSessionListResponse> {
  return apiFetch<PromptEditSessionListResponse>(
    `/projects/${projectId}/chapters/${chapterId}/prompt-edit/sessions`,
  );
}

export async function instructPromptEdit(
  projectId: string,
  chapterId: string,
  body: PromptEditInstructRequest,
): Promise<PromptEditInstructResponse> {
  return apiFetch<PromptEditInstructResponse>(
    `/projects/${projectId}/chapters/${chapterId}/prompt-edit/instruct`,
    {
      method: "POST",
      body: JSON.stringify(body),
    },
  );
}

export async function regeneratePromptEdit(
  projectId: string,
  chapterId: string,
  body: PromptEditRegenerateRequest,
): Promise<PromptEditInstructResponse> {
  return apiFetch<PromptEditInstructResponse>(
    `/projects/${projectId}/chapters/${chapterId}/prompt-edit/regenerate`,
    {
      method: "POST",
      body: JSON.stringify(body),
    },
  );
}

export async function applyPromptEdit(
  projectId: string,
  chapterId: string,
  body: PromptEditApplyRequest,
): Promise<PromptEditApplyResponse> {
  return apiFetch<PromptEditApplyResponse>(
    `/projects/${projectId}/chapters/${chapterId}/prompt-edit/apply`,
    {
      method: "POST",
      body: JSON.stringify(body),
    },
  );
}

File: apps/api/app/routers/prompt_edit.py
"""Prompt edit routes."""

from fastapi import APIRouter, status

from app.deps import ChapterAccess, CurrentUserId, DbSession, ProjectAccess
from app.schemas.prompt_edit import (
    PromptEditApplyRequest,
    PromptEditApplyResponse,
    PromptEditInstructRequest,
    PromptEditInstructResponse,
    PromptEditRegenerateRequest,
    PromptEditSessionListResponse,
)
from app.services.prompt_edit import PromptEditService

router = APIRouter(
    prefix="/projects/{project_id}/chapters/{chapter_id}/prompt-edit",
    tags=["PromptEdit"],
)


@router.post(
    "/instruct",
    response_model=PromptEditInstructResponse,
    status_code=status.HTTP_201_CREATED,
)
async def prompt_edit_instruct(
    payload: PromptEditInstructRequest,
    project: ProjectAccess,
    chapter: ChapterAccess,
    user_id: CurrentUserId,
    session: DbSession,
) -> PromptEditInstructResponse:
    return await PromptEditService(session).instruct(project, chapter, user_id, payload)


@router.post(
    "/regenerate",
    response_model=PromptEditInstructResponse,
    status_code=status.HTTP_201_CREATED,
)
async def prompt_edit_regenerate(
    payload: PromptEditRegenerateRequest,
    project: ProjectAccess,
    chapter: ChapterAccess,
    user_id: CurrentUserId,
    session: DbSession,
) -> PromptEditInstructResponse:
    return await PromptEditService(session).regenerate(project, chapter, user_id, payload)


@router.post("/apply", response_model=PromptEditApplyResponse, status_code=status.HTTP_201_CREATED)
async def prompt_edit_apply(
    payload: PromptEditApplyRequest,
    project: ProjectAccess,
    chapter: ChapterAccess,
    user_id: CurrentUserId,
    session: DbSession,
) -> PromptEditApplyResponse:
    return await PromptEditService(session).apply(project, chapter, user_id, payload)


@router.get("/sessions", response_model=PromptEditSessionListResponse)
async def list_prompt_edit_sessions(
    project: ProjectAccess,
    chapter: ChapterAccess,
    session: DbSession,
) -> PromptEditSessionListResponse:
    return await PromptEditService(session).list_sessions(project, chapter)
